From d86f1aac001174ee92b0075672b12309e20e16c8 Mon Sep 17 00:00:00 2001
From: "Tan, Jia Yong" <jia.yong.tan@intel.com>
Date: Fri, 10 Jan 2025 16:21:49 +0800
Subject: [PATCH] Fix cloud-init selinux permission denied

Audit denied logs:
AVC avc:  denied  { search } for  pid=1151 comm="cloud-init"
name=".mount" dev="sda3" ino=16
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=1
SYSCALL arch=c000003e syscall=137 success=yes exit=0 a0=569b06ac95e6
a1=7fff45b43920 a2=8 a3=7a439ff785e0 items=1 ppid=1 pid=1151
auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0
tty=(none) ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { read } for  pid=1151 comm="cloud-init"
name="enp0s3.lease" dev="overlay" ino=142
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_state_t:s0 tclass=file permissive=1
AVC avc:  denied  { open } for  pid=1151 comm="cloud-init"
path="/var/lib/dhcpcd/enp0s3.lease" dev="overlay" ino=142
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_state_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=257 success=yes exit=5 a0=ffffff9c
a1=7a439e0ae410 a2=80000 a3=0 items=0 ppid=1 pid=1151 auid=4294967295
uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { getattr } for  pid=1151 comm="cloud-init"
path="/var/lib/dhcpcd/enp0s3.lease" dev="overlay" ino=142
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_state_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=262 success=yes exit=0 a0=5
a1=7a439ff9bbd5 a2=7fff45b43c60 a3=1000 items=0 ppid=1 pid=1151
auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0
tty=(none) ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { ioctl } for  pid=1151 comm="cloud-init"
path="/var/lib/dhcpcd/enp0s3.lease" dev="overlay" ino=142
ioctlcmd=0x5401 scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_state_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=16 success=no exit=-25 a0=5 a1=5401
a2=7fff45b43e00 a3=1000 items=0 ppid=1 pid=1151 auid=4294967295 uid=0
gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { read } for  pid=1151 comm="cloud-init"
name="enp0s3-4.pid" dev="tmpfs" ino=1436
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_runtime_t:s0 tclass=file permissive=1
AVC avc:  denied  { open } for  pid=1151 comm="cloud-init"
path="/run/dhcpcd/enp0s3-4.pid" dev="tmpfs" ino=1436
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_runtime_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=257 success=yes exit=5 a0=ffffff9c
a1=7a439e0ae550 a2=80000 a3=0 items=0 ppid=1 pid=1151 auid=4294967295
uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { getattr } for  pid=1151 comm="cloud-init"
path="/run/dhcpcd/enp0s3-4.pid" dev="tmpfs" ino=1436
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_runtime_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=262 success=yes exit=0 a0=5
a1=7a439ff9bbd5 a2=7fff45b43c60 a3=1000 items=0 ppid=1 pid=1151
auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0
tty=(none) ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { ioctl } for  pid=1151 comm="cloud-init"
path="/run/dhcpcd/enp0s3-4.pid" dev="tmpfs" ino=1436 ioctlcmd=0x5401
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:dhcpc_runtime_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=16 success=no exit=-25 a0=5 a1=5401
a2=7fff45b43e00 a3=1000 items=0 ppid=1 pid=1151 auid=4294967295 uid=0
gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { kill } for  pid=1151 comm="cloud-init" capability=5
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:system_r:cloud_init_t:s0 tclass=capability
permissive=1
AVC avc:  denied  { sigkill } for  pid=1151 comm="cloud-init"
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:system_r:dhcpc_t:s0 tclass=process permissive=1
SYSCALL arch=c000003e syscall=62 success=yes exit=0 a0=fffffb66 a1=9
a2=0 a3=7fff45b43c60 items=0 ppid=1 pid=1151 auid=4294967295 uid=0 gid=0
euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295
comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { read } for  pid=1199 comm="cloud-init" name="2"
dev="tmpfs" ino=1194 scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:systemd_networkd_runtime_t:s0 tclass=file
permissive=1
AVC avc:  denied  { open } for  pid=1199 comm="cloud-init"
path="/run/systemd/netif/leases/2" dev="tmpfs" ino=1194
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:systemd_networkd_runtime_t:s0 tclass=file
permissive=1
SYSCALL arch=c000003e syscall=257 success=yes exit=5 a0=ffffff9c
a1=7875e5078510 a2=80000 a3=0 items=0 ppid=1 pid=1199 auid=4294967295
uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { getattr } for  pid=1199 comm="cloud-init"
path="/run/systemd/netif/leases/2" dev="tmpfs" ino=1194
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:systemd_networkd_runtime_t:s0 tclass=file
permissive=1
SYSCALL arch=c000003e syscall=262 success=yes exit=0 a0=5
a1=7875e719bbd5 a2=7ffcd944a5b0 a3=1000 items=0 ppid=1 pid=1199
auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0
tty=(none) ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { ioctl } for  pid=1199 comm="cloud-init"
path="/run/systemd/netif/leases/2" dev="tmpfs" ino=1194 ioctlcmd=0x5401
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:systemd_networkd_runtime_t:s0 tclass=file
permissive=1
SYSCALL arch=c000003e syscall=16 success=no exit=-25 a0=5 a1=5401
a2=7ffcd944a750 a3=1000 items=0 ppid=1 pid=1199 auid=4294967295 uid=0
gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { getattr } for  pid=1199 comm="cloud-init"
path="/usr/share/netplan/netplan.script" dev="sda2" ino=17139
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=262 success=yes exit=0 a0=ffffff9c
a1=7875e574d090 a2=7ffcd944a8f0 a3=0 items=0 ppid=1 pid=1199
auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0
tty=(none) ses=4294967295 comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { execute } for  pid=1199 comm="cloud-init"
name="netplan.script" dev="sda2" ino=17139
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=21 success=yes exit=0 a0=7875e574d090 a1=1
a2=0 a3=2 items=0 ppid=1 pid=1199 auid=4294967295 uid=0 gid=0 euid=0
suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295
comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { read open } for  pid=1214 comm="cloud-init"
path="/usr/share/netplan/netplan.script" dev="sda2" ino=17139
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=file permissive=1
AVC avc:  denied  { execute_no_trans } for  pid=1214 comm="cloud-init"
path="/usr/share/netplan/netplan.script" dev="sda2" ino=17139
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=59 success=yes exit=0 a0=7875e58e7590
a1=7875e58e7f70 a2=7875e508d8e0 a3=8 items=2 ppid=1199 pid=1214
auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0
tty=(none) ses=4294967295 comm="netplan" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { ioctl } for  pid=1214 comm="netplan"
path="/usr/share/netplan/netplan.script" dev="sda2" ino=17139
ioctlcmd=0x5401 scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=16 success=no exit=-25 a0=3 a1=5401
a2=7ffc4151cc30 a3=1000 items=0 ppid=1199 pid=1214 auid=4294967295 uid=0
gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none)
ses=4294967295 comm="netplan" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { unlink } for  pid=1199 comm="cloud-init"
name="ssh_host_rsa_key.pub" dev="sda3" ino=186
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:sshd_key_t:s0 tclass=file permissive=1
SYSCALL arch=c000003e syscall=87 success=yes exit=0 a0=7875e50ccb50 a1=0
a2=0 a3=1 items=0 ppid=1 pid=1199 auid=4294967295 uid=0 gid=0 euid=0
suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295
comm="cloud-init" exe="/usr/bin/python3.12"
subj=system_u:system_r:cloud_init_t:s0 key=(null)
AVC avc:  denied  { read } for  pid=920 comm="cloud-init" name="sda2"
dev="devtmpfs" ino=131 scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:fixed_disk_device_t:s0 tclass=blk_file
permissive=1
AVC avc:  denied  { open } for  pid=920 comm="cloud-init"
path="/dev/sda2" dev="devtmpfs" ino=131
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:fixed_disk_device_t:s0 tclass=blk_file
permissive=1

Audit log [Jan 10 2025]
type=AVC msg=audit(1736487867.571:161): avc:  denied  { map } for
pid=2558 comm="trust"
path="/usr/share/p11-kit/modules/p11-kit-trust.module" dev="dm-5"
ino=17109 scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=file permissive=0
AVC avc:  denied  { sqpoll } for  pid=2857 comm="dig"
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:system_r:cloud_init_t:s0 tclass=io_uring permissive=1
AVC avc:  denied  { create } for  pid=2857 comm="dig"
anonclass=[io_uring] scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:cloud_init_t:s0 tclass=anon_inode
permissive=1
AVC avc:  denied  { map } for  pid=2857 comm="dig"
path="anon_inode:[io_uring]" dev="anon_inodefs" ino=16882
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:cloud_init_t:s0 tclass=anon_inode
permissive=1
AVC avc:  denied  { read write } for  pid=2857 comm="dig"
path="anon_inode:[io_uring]" dev="anon_inodefs" ino=16882
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:cloud_init_t:s0 tclass=anon_inode
permissive=1
AVC avc:  denied  { read } for  pid=1274 comm="trust" name="blocklist"
dev="sda2" ino=6039 scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:usr_t:s0 tclass=lnk_file permissive=1
AVC avc:  denied  { map } for  pid=1284 comm="trust"
path="/etc/pki/ca-trust/source/README" dev="sda3" ino=79
scontext=system_u:system_r:cloud_init_t:s0
tcontext=system_u:object_r:cert_t:s0 tclass=file permissive=1

Signed-off-by: Tan, Jia Yong <jia.yong.tan@intel.com>
---
 policy/modules/admin/cloudinit.te | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/policy/modules/admin/cloudinit.te b/policy/modules/admin/cloudinit.te
index 94f7eabd2..d0d636468 100644
--- a/policy/modules/admin/cloudinit.te
+++ b/policy/modules/admin/cloudinit.te
@@ -2,6 +2,12 @@ policy_module(cloudinit)
 
 gen_require(`
 	class passwd passwd;
+        type fixed_disk_device_t;
+        type dhcpc_t;
+        type dhcpc_state_t;
+        type dhcpc_runtime_t;
+        type systemd_networkd_runtime_t;
+        type unlabeled_t;
 ')
 
 ########################################
@@ -45,13 +51,27 @@ files_tmp_file(cloud_init_tmp_t)
 # Local policy
 #
 
-allow cloud_init_t self:capability { chown dac_override dac_read_search fowner fsetid setgid setuid };
+allow cloud_init_t self:capability { chown dac_override dac_read_search fowner fsetid setgid setuid kill };
 dontaudit cloud_init_t self:capability { net_admin sys_admin sys_tty_config };
 allow cloud_init_t self:fifo_file rw_fifo_file_perms;
 allow cloud_init_t self:unix_dgram_socket create_socket_perms;
 allow cloud_init_t self:passwd passwd;
+allow cloud_init_t self:anon_inode { create map read write };
+allow cloud_init_t self:io_uring sqpoll;
 
 allow cloud_init_t cloud_init_log_t:file { create_file_perms append_file_perms read setattr };
+allow cloud_init_t fixed_disk_device_t:blk_file { read open };
+allow cloud_init_t usr_t:file { execute execute_no_trans getattr ioctl open read map };
+allow cloud_init_t usr_t:lnk_file read;
+allow cloud_init_t cert_t:file map;
+allow cloud_init_t dhcpc_t:process sigkill;
+allow cloud_init_t dhcpc_runtime_t:file { getattr ioctl open read };
+allow cloud_init_t dhcpc_state_t:file { getattr ioctl open read };
+allow cloud_init_t fixed_disk_device_t:blk_file { open read };
+allow cloud_init_t sshd_key_t:file unlink;
+allow cloud_init_t systemd_networkd_runtime_t:dir { getattr write add_name };
+allow cloud_init_t systemd_networkd_runtime_t:file { getattr ioctl write create open read };
+allow cloud_init_t unlabeled_t:dir search;
 logging_log_filetrans(cloud_init_t, cloud_init_log_t, file)
 
 manage_files_pattern(cloud_init_t, cloud_init_runtime_t, cloud_init_runtime_t)
-- 
2.34.1

